docs/docs/shell/self-extractable-archives.mdx

```diff
@@ -0,0 +1,265 @@
+
+```bash
+pants package :arc && dist/arc.run
+```
+```
+[INFO] Wrote dist/arc.run
+Built Makeself binary: arc.run
+Verifying archive integrity...  100%   MD5 checksums are OK. All good.
+Uncompressing arc.run  100%  
+one
+```
+
+To pass the arguments to the startup_script use `--`:
+
+```bash
+pants package :arc && dist/arc.run -- zero
+```
+```
+[INFO] Wrote dist/arc.run
+Built Makeself binary: arc.run
+Verifying archive integrity...  100%   MD5 checksums are OK. All good.
+Uncompressing arc.run  100%  
+zero
+```
+
+## `pants run`
+
+Instead of packaging and running `makeself_archive` manually, you can use `run` goal instead:
+
+```bash
+pants run :arc
+```
+```
+Verifying archive integrity...  100%   MD5 checksums are OK. All good.
+Uncompressing arc.run  100%  
+one
+```
+
+To pass the arguments through the `pants run` goal you need `--`, then you need
+another `--` to pass arguments to archive `startup_script`, so you end up with
+two `--`:
+
+```bash
+pants run :arc -- -- zero
+```
+```
+Verifying archive integrity...  100%   MD5 checksums are OK. All good.
+Uncompressing arc.run  100%  
+zero
+```
+
+Similarly you can pass flags to the archive, for example, `quiet` flag to suppress progress messages:
+```bash
+pants run :arc -- -q -- zero
+```
+```
+zero
+```
+
+## Bundling packages like `pex_binary`
+
+You can put other packages like `pex_binary` into a makeself archive. 
+
+To configure `pex_binary`, first, update your `pants.toml`:
+
+```toml title="pants.toml"
+backend_packages = [
+    ...
+    "pants.backend.shell",
+    "pants.backend.experimental.makeself",
+    "pants.backend.python",
+]
+
+[python]
+interpreter_constraints = ["CPython==3.12.*"]
+```
+
+Now define the `pex_binary` and add it to the makeself archive via the
+`packages` field:
+
+```python tab={"label":"BUILD"}
+python_sources(name="py")
+pex_binary(
+    name="upper",
+    entry_point="upper.py",
+)
+shell_sources(name="sh")
+makeself_archive(
+    name="arc",
+    files=["lib.sh:sh", "entrypoint.sh:sh"],
+    packages=[":upper"],
+    startup_script=["./entrypoint.sh"],
+)
+```
+
+```python tab={"label":"upper.py"}
+print(input().upper())
+```
+
+```bash tab={"label":"entrypoint.sh"} 
+#!/bin/bash
+
+. lib.sh
+echo $@ "one two three" | first_column | ./upper.pex
+```
+
+```bash tab={"label":"lib.sh"} 
+#!/bin/bash
+
+function first_column {
+    awk '{print $1}'
+}
+```
+
+```bash
+pants run :arc -- -q -- zero
+```
+```
+/usr/bin/env: ‘python3.12’: No such file or directory
+```
+
+Oops! This happened because `pants run` is running in isolated environment, so
+we have to explicitly tell pants that we want to access the python interpreter
+and a couple of binaries used by pex:
+
+```python title="BUILD"
+...
+makeself_archive(
+    name="arc",
+    files=["lib.sh:sh", "entrypoint.sh:sh"],
+    packages=[":upper"],
+    startup_script=["./entrypoint.sh"],
+    tools=["python3.12", "grep", "sort"],
```

Comment:
Hm, just observing that these tools seem to be used for running... but I imagine they're not required for packaging? Potentially something to iterate on?

Reply:
This is specifically about the pex inside. Since the `run` goal doesn't know there is a pex inside a makeself archive, you need to explicitly add python, grep and sort which are used inside the pex launcher, but not required for makeself launcher.

This is not needed for shell-only makeself archives.

Reply:
Yeah, I understand it's just for this particular archive... but I think the tools are only used when running the makeself archive using `pants run`? That is, even if this had no tools specified `tools=[]`, I think the `pants package ...` goal would still work.

No need to change, but just observing that there's a distinction between "tools required to package" and "tools required to run", and overlaying them may lead to spurious cache invalidations (etc.).

Reply:
Ah, yes, this is correct

I could split it into `execution_dependencies` and `runnable_dependencies` similar to `adhoc_tool`